crates/core/tedge_api/src/messages.rs

```diff
@@ -40,38 +39,173 @@ where
     }
 }
 
-pub const fn software_filter_topic() -> &'static str {
-    "tedge/commands/req/software/#"
+/// Command to request the list of software packages that are installed on a device
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SoftwareListCommand {
+    pub target: EntityTopicId,
+    pub cmd_id: String,
+    pub payload: SoftwareListCommandPayload,
 }
 
-pub const fn control_filter_topic() -> &'static str {
-    "tedge/commands/req/control/#"
+/// Payload of a [SoftwareListCommand]
+#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct SoftwareListCommandPayload {
+    #[serde(flatten)]
+    pub status: CommandStatus,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub current_software_list: Vec<SoftwareList>,
 }
 
-/// Message payload definition for SoftwareList request.
-#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
-#[serde(deny_unknown_fields)]
+impl<'a> Jsonify<'a> for SoftwareListCommandPayload {}
+
+impl Default for SoftwareListCommandPayload {
+    fn default() -> Self {
+        SoftwareListCommandPayload {
+            status: CommandStatus::Init,
+            current_software_list: Vec::new(),
+        }
+    }
+}
+
+/// Sub list of modules grouped by plugin type.
+#[derive(Debug, Clone, Deserialize, Eq, PartialEq, Serialize)]
 #[serde(rename_all = "camelCase")]
-pub struct SoftwareListRequest {
-    pub id: String,
+#[serde(deny_unknown_fields)]
+pub struct SoftwareList {
+    #[serde(rename = "type")]
+    pub plugin_type: SoftwareType,
+    pub modules: Vec<SoftwareModuleItem>,
 }
 
-impl<'a> Jsonify<'a> for SoftwareListRequest {}
+impl SoftwareListCommand {
+    /// Build a new software list request with a random id
+    pub fn new(target: &EntityTopicId) -> Self {
+        SoftwareListCommand {
+            target: target.clone(),
+            cmd_id: nanoid!(),
+            payload: Default::default(),
+        }
+    }
 
-impl Default for SoftwareListRequest {
-    fn default() -> SoftwareListRequest {
-        let id = nanoid!();
-        SoftwareListRequest { id }
+    /// Build a new software list request with a given id
+    pub fn new_with_id(target: &EntityTopicId, cmd_id: String) -> Self {
```

Comment:
TODO: be consistent with `RestartCommand` constructors.

Reply:
Addresses by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/ecc2a84b7883650a1dadb5bce23e1b127db18872